### ISSUE TITLE
Create an API key with the action `documents.*`

### ISSUE THREAD

**Issue by Mcdostone** [meilisearch/meilisearch#2080]
**Describe the bug**

I would like to create a new API key with restricted permissions on a given index. The API returns an error when I specify the action `documents.*`. According to the [documentation of v0.25](https://docs.meilisearch.com/reference/api/keys.html#actions-2), 
> The * character can also be used as a wildcard to get all permissions for a route, e.g. documents.* to authorize access to all document-related endpoints.

Is it a bug? Is the documentation not up to date?

a simple workaround is to list all actions: `["documents.add", "documents.get", "documents.delete"]`

**To Reproduce**
1.Boot a meilisearch instance
```bash
./meilisearch --master-key key
```
2. Create a new API key
```bash
curl \
  -X POST 'http://localhost:7700/keys' \
  -H 'Authorization: Bearer key' \
  -H 'Content-Type: application/json' \
  --data-binary '{
    "description": "Add documents: Products API key",
    "actions": [
      "documents.*"
    ],
    "indexes": [
    "products"
    ],
    "expiresAt": null
  }'
```
I got the following error:
```json
{
  "message":"actions field value `[\"documents.*\"]` is invalid. It should be an array of string representing action names.",
  "code":"invalid_api_key_actions",
  "type":"invalid_request",
  "link":"https://docs.meilisearch.com/errors#invalid_api_key_actions"
}
```

**Comment by gmourier** [comment]
Hi @Mcdostone 👋

Thanks for this issue, this is a helper we decided to add in a next iteration and I forgot to remove it from the [specification](https://github.com/meilisearch/specifications/blob/main/text/0085-api-keys.md) file used by the documentation team. So this is entirely my fault. Apologize to everyone on that.

@meilisearch/docs-team I've prepared a [pull request](https://github.com/meilisearch/documentation/pull/1370) to remove that specific lines from the documentation until we add it.
@ManyTheFish Is it something we can add for v0.26.0?

**Comment by curquiza** [comment]
Hello @Mcdostone 
Thanks for this report, this is indeed an issue on the docs side currently, this should be fixed with the @gmourier's PR.

I've discussed with @gmourier, this is an issue for the community as "good first issue".
The user expectations are described in [this spec](https://github.com/meilisearch/specifications/blob/main/text/0085-api-keys.md).

**Comment by ManyTheFish** [comment]
Hello! 
For the one who would like to implement this feature, all the changes to make should be in the [meilisearch-auth module](https://github.com/meilisearch/MeiliSearch/tree/main/meilisearch-auth/src). I can give more explanation about how to implement it if someone open a PR.

**Comment by karachikorai** [comment]
@ManyTheFish , If this issue is still open, I can take a look. There are no Pull request for this. Please find the below modifications and let me know if some of these makes sense. I am new to RUST.

Can you please confirm at least some of this modifications are right?

https://github.com/meilisearch/MeiliSearch/blob/main/meilisearch-auth/src/action.rs

pub enum Action {
  #[serde(rename = "documents.*")]
    DocumentsAdd = actions::DOCUMENTS_ADD,
    DocumentsGet = actions::DOCUMENTS_GET,
    DocumentsDelete = actions::DOCUMENTS_DELETE,
}

impl Action {
    RESTRICTED_ACCESS => Some(Self::DocumentsAdd),Some(Self::DocumentsGet),Some(Self::DocumentsDelete)

}

pub fn repr(&self) -> u8 { not able to sort this out}

pub mod actions {
    pub const RESTRICTED_ACCESS: u8 = 16;
}

**Comment by ManyTheFish** [comment]
Hello @karachikorai, welcome!

## Testing
First of all, you should modify the tests, below an hashmap that contains meilisearch routes and the corresponding action to have access to them:
https://github.com/meilisearch/meilisearch/blob/main/meilisearch-http/tests/auth/authorization.rs#L10-L49

This hashmap is used in our test to make sure that implemented actions work and and don't give access to routes that shouldn't be accessed.
Let's modify [theses lines (L13-L16)](https://github.com/meilisearch/meilisearch/blob/main/meilisearch-http/tests/auth/authorization.rs#L13-L16) into:
```diff
-            ("POST",    "/indexes/products/documents") =>                      hashset!{"documents.add", "*"},
-            ("GET",     "/indexes/products/documents") =>                      hashset!{"documents.get", "*"},
-            ("GET",     "/indexes/products/documents/0") =>                    hashset!{"documents.get", "*"},
-            ("DELETE",  "/indexes/products/documents/0") =>                    hashset!{"documents.delete", "*"}, 
+            ("POST",    "/indexes/products/documents") =>                      hashset!{"documents.add", "documents.*", "*"},
+            ("GET",     "/indexes/products/documents") =>                      hashset!{"documents.get", "documents.*", "*"},
+            ("GET",     "/indexes/products/documents/0") =>                    hashset!{"documents.get", "documents.*", "*"},
+            ("DELETE",  "/indexes/products/documents/0") =>                    hashset!{"documents.delete", "documents.*", "*"}, 
```

And make sure that some of the tests (`cargo test auth`) are now failing.

**Comment by karachikorai** [comment]
Thanks @ManyTheFish

**Comment by 3man75** [comment]
Hi all, I'm somewhat familiar with programming and I am interested in helping. This would be my first time ever doing so. 

I am however a little afraid because I am not exactly sure how to tackle the issue. If anyone can provide hints or anything that'd be really nice. If the issue is fixed I'd like to learn more about how the issue was resolved and how it was gone about. Thank you!

**Comment by curquiza** [comment]
Hello @3man75 

The expected implementation and the expected tests are described by Many in [this comment above](https://github.com/meilisearch/meilisearch/issues/2080#issuecomment-1022952486). Is it enough for you? If not, where do you need deeper explanations? 😊 

The expected result in terms of user-experience is described in this [specification part](https://github.com/meilisearch/specifications/blob/main/text/0085-api-keys.md#155-post-patch---keys---api-key-object-payload-definition)

**Comment by janithpet** [comment]
Hi!

I have made an attempt at this in #2525. 

I am super new to Rust; any thoughts / comments regarding this would be welcome! It seems that the PR is failing a Rust formatting check; would you be able to point me to where the issue is?

Thanks!

**Comment by curquiza** [comment]
Hello @janithpet 

Thanks a lot for contributing!!! 😄 

Regarding your formatting issue, the CI failed because the following command is failing (cf [line in the CI manifest](https://github.com/meilisearch/meilisearch/blob/3340af1ba9f82eaf5578b837b52d33aaaac927cc/.github/workflows/rust.yml#L92))

```bash
cargo fmt --all -- --check
```

If I'm correct, if you want to make `cargo fmt` fix the issue, you just have to remove the `--check` option, run the command and commit the changes:

```bash
cargo fmt --all
```

Someone in the core team will review your PR as soon as possible!
For any other questions, feel free to ask directly on the PR

Thanks again for the time you spent on it! ❤️

**Comment by janithpet** [comment]
Thanks @curquiza! 

The tip about `cargo fmt` fixed the issue! The tests now run fine :)

**Comment by curquiza** [comment]
@meilisearch/docs-team

**Comment by curquiza** [comment]
Hello everyone watching this issue!

We have just released [v0.29.0rc1](https://github.com/meilisearch/meilisearch/releases/tag/v0.29.0rc1), which is a release candidate of v0.29.0 🔥 
It means you can test the just-introduced `*` helper when using the `actions` of tenant tokens! [Binaries](https://github.com/meilisearch/meilisearch/releases/tag/v0.29.0rc1) are attached to the release, or you can you the docker image:
```bash
docker run -it --rm \
    -p 7700:7700 \
    getmeili/meilisearch:v0.29.0rc1
```

Let us know about any bugs or feedback! 😄 It would be really helpful.

FYI, the official v0.29.0 release will be available on 3rd October.
